### DISCUSSION TITLE
Conservation in a composite linear solve

### DISCUSSION THREAD

**Discussion by ZiyangHuang** [AMReX-Codes/amrex#3493] [Q&A] [answered]
Greetings,

When I use the composite linear solve (solving all amr levels altogether), do I need to additionally do ```averagedown``` for the solution and ```averagedown_face``` for the fluxes to ensure conservation?

Thanks!

**Comment by WeiqunZhang** [comment, marked as answer]
You don't need to call `averagdown` on the solution. However, you do need to call `averagedown_faces` on fluxes if that's what you want.

**Reply by ZiyangHuang** [reply]
Thanks for your reply. I have a follow-up question. In the example code, the composite solver covers all the amr levels (i.e., from ```lev=0``` to ```lev=max_level```). However, in practice, we may only look for a solution from ```lev=0``` to ```lev=finest_level<=max_level```. Does it mean that it is preferable to have a composite solver covering ```lev=0``` to ```lev=finest_level```. If that's the case, the vectors of ```geom```, ```grids```, and ```dmap``` for building a ```MLLinOP``` should contain data only from ```lev=0``` to ```lev=finest_level```, should they?

**Comment by WeiqunZhang** [comment]
No code was ever mentioned in this Discussion. So I am not sure what's the example code you are talking about. Anyway, the sizes of the vectors of `geom`, `grids` and `dmap` should be the same. For AmrCore and AmrLevel based codes, there are no `grids` and `dmap` for levels above the current finest level (i.e., `AmrMesh::finest_level`). So it's not possible to solve from level 0 to `max_level`, if `max_level` > `finest_level`.

**Reply by ZiyangHuang** [reply]
The example code I mentioned was "ABecLaplacian_C". I am trying to adapt it to Advection_AmrCore to build a convection-diffusion solver. Based on what you said, in AmrCore, ```geom.size() = finest_level```, does it?

Also, it seems not allowed to construct a vector of ```MultiFab``` with ```{}```, which works for ```geom```, ```grids```, and ```dmap```.

**Reply by WeiqunZhang** [reply]
> For AmrCore and AmrLevel based codes, there are no grids and dmap for levels above the current finest level (i.e., AmrMesh::finest_level).

I intentionally did not mention `geom` because its size is `max_level+1`, not necessarily `finest_level+1`. You can use https://github.com/AMReX-Codes/amrex/blob/3396b1df117532ec9a7eafdcaad4bdd18ac49a0f/Src/AmrCore/AMReX_AmrMesh.H#L106 to a vector of Geometry from level 0 to finest_level. Maybe what I said about the sizes of the vectors of geom, grids, dmap should be the same was confusing. What I meant there was the requirement on the arguments passed to the linear solver. At that time, AmrCore had not been mentioned yet.

>  Also, it seems not allowed to construct a vector of MultiFab with {}

Not sure what it means.

**Reply by ZiyangHuang** [reply]
Thanks a lot for your reply. Let me rephase my second question. Let ```MF``` is a defined ```MultiFab```, is it possible to pass ```{MF}``` as a vector of ```MultiFab``` to a function?

**Reply by WeiqunZhang** [reply]
No, because MultiFab does not have copy constructor. This is intentional because we do not want to unnecessarily deep-copy a MultiFab. That's why the interface of a lot of amrex functions uses vector of multifab pointers.

(Could one store MulitFab in a vector? Yes, because MultiFab has a move constructor and move assignment operator.)

**Reply by ZiyangHuang** [reply]
Thanks! It is clear to me now.
